### PR TITLE
[shopsys] performance improvement on product list/edit

### PR DESCRIPTION
**Pull request by grossmannmartin**
| Q                                                                                                              | A
|----------------------------------------------------------------------------------------------------------------| ---
| Description, the reason for the PR                                                                             | Minor performance improvements for product list and product edit page in administration. The performance impact is even more visible on e-shop with more domains.
| New feature                                                                                                    | No 
| [BC breaks](https://docs.shopsys.com/en/latest/contributing/backward-compatibility-promise/)                   | No 
| Fixes issues                                                                                                   | ... 
| Have you read and signed our [License Agreement for contributions](https://www.shopsys.com/license-agreement)? | Yes

DB query count comparison on demo data with two domains

| Page         |  DB query count before | DB query count after |
|--------------|--------------|-------------|
| Product Grid | 138          | 107         |
| Product Hello kitty Edit | 196          | 138         |

----
:globe_with_meridians: Live Preview:
  - https://mg-admin-performance.odin.shopsys.cloud
  - https://cz.mg-admin-performance.odin.shopsys.cloud

### CODE REVIEW

File: packages/framework/src/Model/Product/ProductVisibilityRepository.php [hunk 1]
```diff
@@ -107,6 +107,29 @@ public function findProductVisibilitiesByDomainIdAndProduct($domainId, Product $
...
+    public function getCountOfDomainsProductIsVisibleOn(
+        Product $product,
+        array $defaultPricingGroupIdsIndexedByDomainId,
+    ): int {
+        $queryBuilder = $this->em->createQueryBuilder()
+            ->select('count(pv.domainId)')
+            ->from(ProductVisibility::class, 'pv')
+            ->where('pv.product = :product')
+            ->andWhere('pv.pricingGroup IN (:pricingGroupIds)')
+            ->andWhere('pv.domainId IN (:domainIds)')
```
Comment (vitek-rostislav) [inline packages/framework/src/Model/Product/ProductVisibilityRepository.php]: I wanted to point out that this condition seems to be unnecessary as pricing group is already bound to a domain, however, that could open a discussion about `ProductVisibility::domainId` as a whole so I am not sure :smile:

File: packages/framework/src/Model/Product/Parameter/ParameterRepository.php [hunk 2]
```diff
@@ -125,6 +125,22 @@ public function getAll(): array
...
+    {
+        $this->em->createQueryBuilder()
+            ->select('p', 'pt')
+            ->from(Parameter::class, 'p')
+            ->join('p.translations', 'pt')
+            ->orderBy('p.id', 'asc')
+            ->getQuery()
+            ->execute();
+
+        return $this->getParameterRepository()->findBy([], ['id' => 'asc']);
```
Comment (vitek-rostislav) [inline packages/framework/src/Model/Product/Parameter/ParameterRepository.php]: wow, this is an interesting approach. Executing the query loads the translations so they are then loaded in the following query? Do I understand it correctly? Is not it too magical? :smile: personally, I would prefer something like this - https://github.com/shopsys/shopsys/blob/15.0/packages/framework/src/Model/Payment/PaymentRepository.php#L154

Comment (grossmannmartin) [reply packages/framework/src/Model/Product/Parameter/ParameterRepository.php]: 🙈  this is probably a mistake... I have changed it in bfd03e5c14

File: packages/framework/src/Model/Product/ProductVisibilityRepository.php (hunk 1 above)
Comment (grossmannmartin) [reply packages/framework/src/Model/Product/ProductVisibilityRepository.php]: I have removed the domainId condition.
I kept the domainId as a keys in the array returned from `\Shopsys\FrameworkBundle\Model\Pricing\Group\PricingGroupSettingFacade::getAllDefaultPricingGroupsIdsIndexedByDomainId`
It seems more usable this way
d486350321

File: packages/framework/src/Model/Product/Parameter/ParameterRepository.php [hunk 3]
```diff
@@ -130,15 +127,22 @@ public function getAll(): array
...
     public function getAllWithTranslations(): array
     {
-        $this->em->createQueryBuilder()
-            ->select('p', 'pt')
-            ->from(Parameter::class, 'p')
+        return $this->getAllQueryBuilder()
+            ->addSelect('pt')
             ->join('p.translations', 'pt')
-            ->orderBy('p.id', 'asc')
+            ->orderBy('pt.name', 'asc')
```
Comment (vitek-rostislav) [inline packages/framework/src/Model/Product/Parameter/ParameterRepository.php]: we should use `OrderByCollationHelper` here, right? :innocent:

**Review by vitek-rostislav — APPROVED** [review]
one more thing :slightly_smiling_face:

File: packages/framework/src/Model/Product/Parameter/ParameterRepository.php (hunk 3 above)
Comment (grossmannmartin) [reply packages/framework/src/Model/Product/Parameter/ParameterRepository.php]: exactly


fixed in 77f3bfccb5

File: packages/framework/src/Form/Admin/Product/Parameter/ProductParameterValueFormType.php [hunk 4]
```diff
@@ -21,9 +22,12 @@ class ProductParameterValueFormType extends AbstractType
 
     /**
      * @param \Shopsys\FrameworkBundle\Model\Product\Parameter\ParameterFacade $parameterFacade
+     * @param \Shopsys\FrameworkBundle\Model\Localization\Localization $localization
      */
-    public function __construct(private readonly ParameterFacade $parameterFacade)
-    {
+    public function __construct(
+        private readonly ParameterFacade $parameterFacade,
+        private readonly Localization $localization,
```
Comment (pk16011990) [inline packages/framework/src/Form/Admin/Product/Parameter/ProductParameterValueFormType.php:29]: Missing an upgrade notes?

Comment (grossmannmartin) [reply packages/framework/src/Form/Admin/Product/Parameter/ProductParameterValueFormType.php:29]: the FormTypes are not meant to be extended via inheritance

**Pull request merged**